### PR TITLE
Basic AI movement + MonsterBuilder

### PR DESCRIPTION
**Pull request by rincew1nd**
Here is wonky AI movement + monster builder.

### CODE REVIEW

File: src/ChannelServer/Network/ChannelPacketSender.cs [hunk 1]
```diff
@@ -1472,32 +1472,57 @@ public static void ZC_MOVE_DIR(Character character, float x, float y, float z, f
...
+        /// <param name="monster"></param>
+        /// <param name="x"></param>
+        /// <param name="y"></param>
+        /// <param name="z"></param>
+        public static void ZC_MOVE_PATH(Monster monster, float x, float y, float z)
+        {
+            var packet = new Packet(Op.ZC_MOVE_PATH);
+
+            packet.PutInt(monster.Handle);
+            packet.PutFloat(monster.Position.X);
```
Comment (Tachiorz) [inline src/ChannelServer/Network/ChannelPacketSender.cs]: There should be ints not floats

Comment (rincew1nd) [inline src/ChannelServer/Network/ChannelPacketSender.cs]: ZC_MOVE_PATH contains move vector with coordinates described by int variables, not float. Or am i wrong? Note: Also it contains 2 unknown float after path (first seems to be speed, and second always zero)
Prove:
http://i.imgur.com/iGkvGYD.png
http://mabinogi.pp.ua/g/8d127679d00830f32385bceca3229616.png

File: src/ChannelServer/World/MonsterBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+using Melia.Shared.World;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Melia.Channel.World
+{
+    class MonsterBuilder
```
Comment (exectails) [inline src/ChannelServer/World/MonsterBuilder.cs]: What are you trying to achieve with this class? Why not simply use the Monster class?

File: src/ChannelServer/Network/ChannelPacketSender.cs [hunk 3]
```diff
@@ -432,24 +432,24 @@ public static void ZC_ABILITY_LIST(Character character)
...
+			packet.PutFloat(entity.Speed);
 			packet.PutFloat(0);
 
 			// [i11257 (2016-03-25)]
 			{
 				packet.PutByte(0);
 			}
 
-			character.Map.Broadcast(packet, character);
+            entity.Map.Broadcast(packet, entity);
```
Comment (exectails) [inline src/ChannelServer/Network/ChannelPacketSender.cs]: Please set your IDE to use tabs, as outlined in our coding conventions, which can be found in the [readme](https://github.com/aura-project/melia#coding-conventions).

File: src/ChannelServer/World/MonsterBuilder.cs (hunk 2 above)
Comment (rincew1nd) [inline src/ChannelServer/World/MonsterBuilder.cs]: Builders are easier to use. The way it's done now need from you to specify all major values manually to object, but builders will done it by themselfs with better visualization.

File: src/ChannelServer/Network/ChannelPacketSender.cs (hunk 3 above)
Comment (rincew1nd) [inline src/ChannelServer/Network/ChannelPacketSender.cs]: My bad. Something goes wrong while copy pasting.

File: src/ChannelServer/World/MonsterBuilder.cs (hunk 2 above)
Comment (exectails) [inline src/ChannelServer/World/MonsterBuilder.cs]: That is highly debatable in my opinion. Personally I dislike factories/builders and I don't see the appeal. What you're doing here can be done with the actual object just as easy, and constructors were *invented* to initialize classes with already available information. It's like you're moving the constructor to its own class. Not only that, you're basically creating a copy of `Monster`, forcing us to maintain one class for the price of two.

I think this is terribly inefficient.

File: src/ChannelServer/Scripting/ScriptManager.cs [hunk 4]
```diff
@@ -413,7 +413,7 @@ private int logdebug(IntPtr L)
 		/// <remarks>
 		/// Parameters:
 		/// - int monsterId
-		/// - string name / Localkey
```
Comment (exectails) [inline src/ChannelServer/Scripting/ScriptManager.cs]: Why did you remove this? It documents that you can use either a name or a localization key.

File: src/ChannelServer/Scripting/ScriptManager.cs [hunk 5]
```diff
@@ -443,11 +443,13 @@ private int addnpc(IntPtr L)
 			if (name.StartsWith("ETC_") || name.StartsWith("QUEST_"))
 				name = "@dicID_^*$" + name + "$*^";
 
-			var monster = new Monster(monsterId, NpcType.NPC);
```
Comment (exectails) [inline src/ChannelServer/Scripting/ScriptManager.cs]: There's nothing wrong with the way you initiate monster here and further down, but you might want to refrain from changing things just so they're to *your* liking. It creates unnecessary diffs, and this being an open-source project not everything will always be like you would normally do it.

File: src/ChannelServer/Util/GmCommands.cs [hunk 6]
```diff
@@ -42,7 +42,7 @@ public GmCommands()
 			Add("jump", "<x> <y> <z>", HandleJump);
 			Add("warp", "<map id> <x> <y> <z>", HandleWarp);
 			Add("item", "<item id> [amount]", HandleItem);
-			Add("spawn", "<monster id>", HandleSpawn);
+			Add("spawn", "<monster id> <x> <y> <z>", HandleSpawn);
```
Comment (exectails) [inline src/ChannelServer/Util/GmCommands.cs]: What do you believe we can achieve by letting users specify a position?

The command manager currently can only handle indexed parameters, which makes them a rare commodity, because each new one has an impact on the usability and complicates the parsing. I'd suppose there aren't many who have a need to spawn a monster at a specific position, usually you'd want to spawn it where *you* are, or at the position of someone else, both of which can be achieved without a position. In my opinion there are better uses for the parameters.

On a side-note, the parameters you added are optional, you should use squared brackets to document that, see the item command for an example.

File: src/ChannelServer/World/Entity.cs [hunk 7]
```diff
@@ -10,5 +10,7 @@ public interface IEntity
 		int Hp { get; }
 		Map Map { get; }
 		Position Position { get; }
+		Direction Direction { get; }
+		float Speed { get; }
```
Comment (exectails) [inline src/ChannelServer/World/Entity.cs]: My original thought was to use Speed for the base speed of an entity, while GetSpeed, like in Character provides the actual Speed, factoring in buffs and the like. (A separate method to indicate that something is calculated on the fly.) If you change that, you should remove the GetSpeed method entirely.

File: src/ChannelServer/World/Monster.cs [hunk 8]
```diff
@@ -149,5 +228,157 @@ public void Kill(Character killer)
...
+		
+		/// <summary>
+		/// Sets direction toward target and updates clients.
+		/// </summary>
+		public void RotateTowardTarget()
+		{
+			if (this.Hp <= 0) return;
+
+			if (this.Target == null) return;
+			this.SetDirection(
```
Comment (exectails) [inline src/ChannelServer/World/Monster.cs]: Feel free not to break non-comment lines at 80, as outlined in our coding conventions.

File: src/ChannelServer/World/Monster.cs [hunk 9]
```diff
@@ -149,5 +228,157 @@ public void Kill(Character killer)
...
+				Shared.Util.Math2.AngleBetweenTwoEntity(
```
Comment (exectails) [inline src/ChannelServer/World/Monster.cs]: Note: Personally I believe a `using` increases the readability significantly, because you don't need the full name. Even if it's only for one two instances.

File: src/ChannelServer/World/Monster.cs [hunk 10]
```diff
@@ -149,5 +228,157 @@ public void Kill(Character killer)
...
+
+		/// <summary>
+		/// Attack target entity if monster is agred
+		/// </summary>
+		public void AttackTarget()
+		{
+			if (this.Target != null && this.Hp > 0 && this.IsInRangeOfAttack)
+			{
+				//this.Target.Hp -= this.Attack;
+				RotateTowardTarget();
```
Comment (exectails) [inline src/ChannelServer/World/Monster.cs]: Please use `this.`, as outlined in our coding conventions.

File: src/Shared/Util/Math2.cs [hunk 11]
```diff
@@ -70,5 +73,45 @@ public static bool Between(int val, int min, int max)
...
 			return (val >= min && val <= max);
 		}
+
+		/// <summary>
+		/// Calculate rotation toward target
+		/// </summary>
+		/// <param name="source"></param>
+		/// <param name="target"></param>
+		/// <returns></returns>
+		public static Direction AngleBetweenTwoEntity(Position source, Position target)
```
Comment (exectails) [inline src/Shared/Util/Math2.cs]: I'd propose to move this to Position or Direction, as it's not a general use Math function, but something very specific to the game. Same for VectorTwoEntity.

**Comment by rincew1nd** [comment]
Right now there is no point in changing Speed variable and GetSpeed methods in class.
@gastonb has make some classes for movable NPC, i believe. Need to wait a little before he will merge it to original repository.

**Comment by exectails** [comment]
>Right now there is no point in changing Speed variable and GetSpeed methods in class.

There's no need to change speed? I use the speed command constantly, and you don't know when someone might want to add buffs and the like^^ My point was, it was there and you removed it only partially, which might cause confusion about the difference between Speed and GetSpeed.

**Comment by rincew1nd** [comment]
I removed Speed from IEntity, cause not all entity can move at all. For that reason i added speed variable and method to monster class.
For movable entity we need to have another interface with have speed, @gastonb done this in his repository.
